doc/developer/design/20221204_with_mutually_recursive.md

```diff
@@ -0,0 +1,183 @@
+
+    We must support the `WITH MUTUALLY RECURSIVE` fragment and capture the column names and types.
+    We are in a position to allow and capture further constraint information, such as nullability and unique keys.
+    This happens by default with `parse_columns`, though we could intentionally avoid accepting that information until we know what to do with it.
+
+2.  SQL name resolution
+
+    SQL name resolution currently introduces CTE bindings after resolving names in the CTE.
+    We would need to generalize this to introduce the names before name resolving recursive CTEs.
+
+3.  SQL planning
+
+    SQL planning currently introduces CTE bindings (types mainly) after planning the CTE.
+    We would need to generalize this to introduce the types before planning recursive CTEs. 
+
+4.  HIR generalization
+
+    HIR has a single-binding `Let` variant that cannot house mutually recursive bindings.
+    We could introduce a `LetRec` or `MultiLet` variant to support this, and even migrate other uses to it.
+
+5.  Lowering / decorrelation
+
+    It is not clear that lowering and decorrelation are complicated.
+    As existing single-binding `Let` constructs, the `value` and `body` are each prefixed with correlated columns of the outer scope, and this should continue to be the case.
+    
+6.  MIR generalization
+
+    MIR also requires a multi-binding `Let`, which could be prototyped as `LetRec` or `MultiLet`.
+
+7.  MIR optimization corrections
+
+    All MIR optimizations are at risk of being incorrect with recursive `Let` bindings.
+    Many of them propagate information about let bindings and must either a. start with conservative information (e.g. all columns required) or b. iterate to fixed point before applying their conclusions.
+
+8.  LIR generalization
+
+    LIR may want to *remove* `Let` and move all bindings in to the sequence of dataflow objects to construct.
+    This is already roughly the same as having a multiple binding `Let` block around the whole statement.
+
+9.  Rendering
+
+    Rendering will need a few modifications, but they have been prototyped and seen to work multiple times.
+    The main change is having each bindings result in a differential dataflow `Variable` which is then used instead of the collection itself.
+    Once the definition of each object is complete, it is bound to its variable.
+
+## Alternatives
+
+<!--
+// Similar to the Description section. List of alternative approaches considered, pros/cons or why they were not chosen
+-->
+
+The only known alternative is SQL's `WITH RECURSIVE` which is sufficiently fraught that we ought not implement it directly.
+
+There are several distinctions it has from the proposed `WITH MUTUALLY RECURSIVE` that it may help to visit.
+
+1.  SQL's `WITH RECURSIVE` does not require column names and types.
+
+    In principle the names and types could be inferred.
+    In practice this appears to be a complicated task that would require substantial reconsideration.
```

Comment:
I don't think it's complicated for `WITH RECURSIVE`, actually! At least, in PostgreSQL, the following does not infer the type of `'1'` as `integer`, as PostgreSQL would in other situations:

```
WITH RECURSIVE t(n) AS (
    VALUES ('1')
  UNION ALL
    SELECT n+1 FROM t WHERE n < 100
)
SELECT sum(n) FROM t;
```

Put another way: it looks like the types are entirely determined by the base case, and since the base case cannot refer to any other names, type inference is no harder than it is in a normal query.

Reply:
That's an interesting point, and a reasonable explanation for why the base case is required!